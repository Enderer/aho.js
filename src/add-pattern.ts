import { getChild, Node, Root } from './trie';

export interface Pattern<T = string>
    extends Iterable<T> {};

export interface PatternAsync<T =string>
    extends AsyncIterable<T>{};

/**
 * Add a pattern to the search trie
 * @param root Root node of an aho-corasick trie
 * @param pattern Pattern that should be matched
 */
export const addPattern = <T = string>(root: Root<T>, pattern: Pattern<T>): Root<T> => {
  if (root == null) { throw new Error('Root is null'); }
  if (pattern == null) { return root; }

  let node: Node<T> = root;
  for (const token of pattern) {
    let child = getChild(node, token);
    if (child == null) {
      node.children = node.children ?? [];
      child = {
        token,
        parent: node,
        path: node.path + String(token),
        depth: node.depth + 1
      };
      node.children.push(child);
    }
    node = child;
  }
  node.pattern = pattern;
  return root;
};
