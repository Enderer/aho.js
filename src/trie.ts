import { Pattern } from './add-pattern';

export interface Node<T = string> {
  pattern?: Pattern<T>;
  children?: Node<T>[];
  token?: T;
  parent?: Node<T>;
  fallback?: Node<T>;
  path?: string;
  depth?: number;
}

export interface Root<T = string> extends Node<T> {}

export const createRoot = <T = string>(): Root<T> => ({ path: '^' });

/**
 * Get the child node that matches a given token value if one exists.
 * If no child has the given token return undefined.
 */
export const getChild = <T = string>(node: Node<T>, token: T): Node<T> | undefined => {
  const children = node?.children;
  if (children == null) { return undefined; }
  const child = children.find(c => c.token === token);
  return child;
};

/** Traverse the trie using bread first search */
export function * walkBFS<T> (n: Node<T>): Iterable<Node<T>> {
  const q = [n];
  while (q.length > 0) {
    const node = q.shift();
    if (node != null) { yield node; }

    const children = node?.children ?? [];
    q.push(...children);
  }
}

/** Traverse the trie following fallback nodes until you get to root */
export function * walkFallback<T> (n?: Node<T>): Iterable<Node<T>> {
  let node: Node<T> | undefined = n;
  while (node != null) {
    yield (node);
    node = node.fallback === node ? undefined : node.fallback;
  }
}
