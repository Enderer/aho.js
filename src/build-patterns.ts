import { getChild, Root, walkBFS, walkFallback } from './trie';

/**
 * Once all patterns have been added to the trie prepare it for searching.
 * Creates fallback links for each node using BFS to allow moving to the next
 * pattern when a match fails.
 * @param root Root node of a search trie
 */
export const buildPatterns = <T = string>(root: Root<T>): Root<T> => {
  root.fallback = root;
  for (const node of walkBFS(root)) {
    if (node === root) { continue; }

    for (const n of walkFallback(node.parent?.fallback)) {
      const child = getChild(n, node.token!);
      if (child != null) {
        node.fallback = child;
        break;
      }
    }

    if (node.fallback == null) { node.fallback = root; }
    if (node.fallback === node) { node.fallback = root; }
  }
  return root;
};
