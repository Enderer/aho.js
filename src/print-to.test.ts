import { addPattern } from './add-pattern';
import { buildPatterns } from './build-patterns';
import { printToStream, printToString } from './print-to';
import { createRoot, Root } from './trie';

describe('print-to', () => {
  describe('printToString', () => {
    it('should handle null root', () => {
      expect(printToString(null as unknown as Root)).toEqual('');
    });

    it('should print a trie', () => {
      const patterns = [
        ['he', 'she', 'his', 'hers'],
        ['abcd', 'abcdef', 'abcdefz', 'ab', 'bc', 'bd', 'cd', 'cde', 'e', 'f', 'c'],
        ['antioxidant', 'antipathy', 'antivirus', 'antigen']
      ];
      const tries = patterns
        .map(t => t.reduce((r, p) => addPattern(r, p), createRoot()))
        .map(t => buildPatterns(t));

      tries.forEach(t => expect(printToString(t)).toMatchSnapshot());
    });
  });

  describe('printToStream', () => {
    it('should print a trie', async () => {
      const patterns = [
        ['he', 'she', 'his', 'hers'],
        ['abcd', 'abcdef', 'abcdefz', 'ab', 'bc', 'bd', 'cd', 'cde', 'e', 'f', 'c'],
        ['antioxidant', 'antipathy', 'antivirus', 'antigen']
      ];
      const tries = patterns
        .map(t => t.reduce((r, p) => addPattern(r, p), createRoot()))
        .map(t => buildPatterns(t));

      const strings = tries.map(async t => {
        const readable = printToStream(t);
        let result = '';
        for await (const chunk of readable) {
          result += chunk;
        }
        return result;
      });
      for await (const s of strings) {
        expect(s).toMatchSnapshot();
      }
    });
  });
});
