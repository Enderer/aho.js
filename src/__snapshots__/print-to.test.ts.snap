// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`print-to printToStream should print a trie 1`] = `
"[] ^
  [h] ^
    [e]* ^
      [r] ^
        [s]* ^s
    [i] ^
      [s]* ^s
  [s] ^
    [h] ^h
      [e]* ^he
"
`;

exports[`print-to printToStream should print a trie 2`] = `
"[] ^
  [a] ^
    [b]* ^b
      [c] ^bc
        [d]* ^cd
          [e] ^cde
            [f]* ^f
              [z]* ^
  [b] ^
    [c]* ^c
    [d]* ^
  [c]* ^
    [d]* ^
      [e]* ^e
  [e]* ^
  [f]* ^
"
`;

exports[`print-to printToStream should print a trie 3`] = `
"[] ^
  [a] ^
    [n] ^
      [t] ^
        [i] ^
          [o] ^
            [x] ^
              [i] ^
                [d] ^
                  [a] ^a
                    [n] ^an
                      [t]* ^ant
          [p] ^
            [a] ^a
              [t] ^
                [h] ^
                  [y]* ^
          [v] ^
            [i] ^
              [r] ^
                [u] ^
                  [s]* ^
          [g] ^
            [e] ^
              [n]* ^
"
`;

exports[`print-to printToString should print a trie 1`] = `
"[] ^
  [h] ^
    [e]* ^
      [r] ^
        [s]* ^s
    [i] ^
      [s]* ^s
  [s] ^
    [h] ^h
      [e]* ^he
"
`;

exports[`print-to printToString should print a trie 2`] = `
"[] ^
  [a] ^
    [b]* ^b
      [c] ^bc
        [d]* ^cd
          [e] ^cde
            [f]* ^f
              [z]* ^
  [b] ^
    [c]* ^c
    [d]* ^
  [c]* ^
    [d]* ^
      [e]* ^e
  [e]* ^
  [f]* ^
"
`;

exports[`print-to printToString should print a trie 3`] = `
"[] ^
  [a] ^
    [n] ^
      [t] ^
        [i] ^
          [o] ^
            [x] ^
              [i] ^
                [d] ^
                  [a] ^a
                    [n] ^an
                      [t]* ^ant
          [p] ^
            [a] ^a
              [t] ^
                [h] ^
                  [y]* ^
          [v] ^
            [i] ^
              [r] ^
                [u] ^
                  [s]* ^
          [g] ^
            [e] ^
              [n]* ^
"
`;
